optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -0,0 +1,480 @@
+        return new Constraint[] {
+                oneShiftPerDay(constraintFactory),
+                minimumAndMaximumNumberOfAssignments(constraintFactory),
+                minimumNumberOfAssignmentsNoAssignments(constraintFactory),
+                minimumConsecutiveWorkingDays(constraintFactory),
+                maximumConsecutiveWorkingDays(constraintFactory),
+                minimumConsecutiveFreeDays(constraintFactory),
+                maximumConsecutiveFreeDays(constraintFactory),
+                maximumConsecutiveFreeDaysNoAssignments(constraintFactory),
+                consecutiveFreeDaysFirstBreak(constraintFactory),
+                consecutiveFreeDaysFinalBreak(constraintFactory),
+                minimumConsecutiveWorkingWeekends(constraintFactory),
+                maximumConsecutiveWorkingWeekends(constraintFactory),
+                startOnNotFirstDayOfWeekend(constraintFactory),
+                endOnNotLastDayOfWeekend(constraintFactory),
+                identicalShiftTypesDuringWeekend(constraintFactory),
+                dayOffRequest(constraintFactory),
+                dayOnRequest(constraintFactory),
+                shiftOffRequest(constraintFactory),
+                shiftOnRequest(constraintFactory),
+                alternativeSkill(constraintFactory),
+                unwantedPatternFreeBefore2DaysWithAWorkDayPattern(constraintFactory),
+                unwantedPatternShiftType2DaysPattern(constraintFactory),
+                unwantedPatternShiftType3DaysPattern(constraintFactory),
+        };
+    }
+
+    // ############################################################################
+    // Hard constraints
+    // ############################################################################
+
+    // A nurse can only work one shift per day, i.e. no two shift can be assigned to the same nurse on a day.
+    Constraint oneShiftPerDay(ConstraintFactory constraintFactory) {
+        return constraintFactory
+                .fromUniquePair(ShiftAssignment.class,
+                        Joiners.equal(ShiftAssignment::getEmployee),
+                        Joiners.equal(ShiftAssignment::getShiftDate))
+                .penalize("org.optaplanner.examples.nurserostering.solver", "oneShiftPerDay", HardSoftScore.ONE_HARD);
+    }
+
+    // ############################################################################
+    // Soft constraints
+    // ############################################################################
+    Constraint minimumAndMaximumNumberOfAssignments(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(MinMaxContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine.getContractLineType() == ContractLineType.TOTAL_ASSIGNMENTS &&
+                        minMaxContractLine.isEnabled())
+                .join(Employee.class,
+                        Joiners.equal(MinMaxContractLine::getContract, Employee::getContract))
```

Comment:
It would seem that you do not need the join with the `MinMaxContractLine`.
The same could possibly be done with `ifExists(...)` after `from(Employee.class)`.
This has the potential to speed up things quite a bit.